File: sdks/nodejs/src/utilities/mcp.js
import axios from 'axios';
import { v4 as uuidv4 } from 'uuid';

/**
 * MCP client for the `remote-mcp-tool` node.
 *
 * Speaks JSON-RPC 2.0 over the Streamable HTTP transport defined by
 * the Model Context Protocol spec:
 *
 *   https://modelcontextprotocol.io/specification
 *
 * Spec-relevant wire requirements (all enforced by reference servers
 * like @modelcontextprotocol/sdk's StreamableHTTPServerTransport):
 *
 *   1. Every request body carries `jsonrpc: "2.0"` plus `id` /
 *      `method` / `params` — the JSON-RPC envelope, not just the
 *      params.
 *   2. `Accept` header MUST list both `application/json` AND
 *      `text/event-stream` — the server picks the response shape
 *      based on whether it's streaming.
 *   3. The response body is either:
 *        - `Content-Type: application/json` — single JSON-RPC
 *          response object, parse normally.
 *        - `Content-Type: text/event-stream` — SSE wire shape;
 *          last `data:` line carries the JSON-RPC response.
 *      Both shapes are valid; this client handles either.
 */

/** Build the headers for an MCP request. Adds bearer auth when
 *  provided; sets the Accept header to satisfy spec-compliant
 *  servers that require both content types in the list. */
function buildHeaders(token) {
  const headers = {
    'Content-Type': 'application/json',
    Accept: 'application/json, text/event-stream',
  };
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }
  return headers;
}

/** Parse an MCP HTTP response. Tolerates either content-type
 *  variant (plain JSON or SSE-framed). Returns the JSON-RPC
 *  response object's `.result` (or throws on `.error`). */
function parseMcpResponse(response) {
  const contentType = (response.headers?.['content-type'] || '').toLowerCase();

  // Streamable HTTP servers default to SSE — body is one or more
  // `event: ...\ndata: ...\n\n` blocks. Find the final `data:` line
  // and JSON.parse it. We disabled axios' default JSON parse so
  // `response.data` is the raw text either way.
  if (contentType.includes('text/event-stream')) {
    const raw = typeof response.data === 'string'
      ? response.data
      : String(response.data ?? '');
    const dataLines = [];
    for (const line of raw.split(/\r?\n/)) {
      if (line.startsWith('data:')) dataLines.push(line.slice(5).trimStart());
    }
    if (dataLines.length === 0) {
      throw new Error('MCP SSE response had no data lines');
    }
    const payload = JSON.parse(dataLines[dataLines.length - 1]);
    if (payload.error) {
      throw new Error(
        `MCP error ${payload.error.code}: ${payload.error.message}`,
      );
    }
    return payload.result;
  }

  // Plain JSON response — single JSON-RPC object. `response.data`
  // is the raw text (we disabled axios' parse); JSON.parse here.
  const raw = typeof response.data === 'string'
    ? response.data
    : JSON.stringify(response.data);
  const payload = JSON.parse(raw);
  if (payload?.error) {
    throw new Error(
      `MCP error ${payload.error.code}: ${payload.error.message}`,
    );
  }
  return payload?.result;
}

/**
 * Call an MCP tool.
 *
 * Tool identity and tool arguments are carried as SEPARATE fields —
 * never merged into one flat object. The previous flat-object contract
 * ({ ...toolArgs, name: toolName }) made a tool argument named `name`
 * indistinguishable from the tool's own name: it was overwritten at
 * the call site, then stripped here, so tools with a top-level `name`
 * parameter received `name: undefined`. Any unexpected top-level key
 * throws so a legacy-shape caller fails loudly instead of silently
 * dropping arguments.
 *
 * @param {Object} call - { name: string, arguments?: Object } — the
 *   tool's name and the model-provided arguments, forwarded verbatim.
 * @param {Object} options - { url, token }
 * @returns {Object} The result of the tool call
 */
export async function callMCPTool(call, { url, token } = {}) {
  if (!url) throw new Error('No MCP URL provided');

  const { name: toolName, arguments: toolArgs } = call ?? {};
  if (!toolName) throw new Error('No tool name provided for MCP call');

  const extraKeys = Object.keys(call).filter(
    (k) => k !== 'name' && k !== 'arguments',
  );
  if (extraKeys.length > 0) {
    throw new Error(
      `callMCPTool takes { name, arguments } — unexpected top-level keys: ` +
      `${extraKeys.join(', ')}. Tool arguments belong under "arguments".`,
    );
  }

  const id = uuidv4();

  try {
    const response = await axios.post(
      url,
      {
        jsonrpc: '2.0',
        id,
        method: 'tools/call',
        params: {
          name: toolName,
          arguments: toolArgs ?? {},
        },
      },
      {
        headers: buildHeaders(token),
        // SSE responses arrive as text; tell axios not to JSON.parse.
        responseType: 'text',
        transformResponse: [(d) => d],
      },
    );
    return parseMcpResponse(response);
  } catch (err) {
    throw new Error(
      `Failed to call MCP tool "${toolName}" at ${url}: ${err.message}`,
    );
  }
}

/**
 * Fetch all tool schemas from an MCP endpoint.
 * @param {string} url - The MCP server URL
 * @param {string} [token] - Optional bearer token for authentication
 * @returns {Array} Array of tool schemas (normalized to {name, description, parameters})
 */
export async function fetchMCPTools(url, token) {
  if (!url) throw new Error('No MCP URL provided');

  const id = uuidv4();

  try {
    const response = await axios.post(
      url,
      {
        jsonrpc: '2.0',
        id,
        method: 'tools/list',
        params: {},
      },
      {
        headers: buildHeaders(token),
        responseType: 'text',
        transformResponse: [(d) => d],
      },
    );
    const result = parseMcpResponse(response);
    const tools = result?.tools || [];
    return tools.map((tool) => ({
      name: tool.name,
      description: tool.description,
      parameters: tool.inputSchema,
    }));
  } catch (err) {
    throw new Error(`Failed to fetch MCP tools from ${url}: ${err.message}`);
  }
}

File: sdks/python/src/mcp.py
"""
MCP (Model Context Protocol) utilities for the zv1 engine.

Provides functions for calling remote MCP tools and fetching tool schemas.
MCP is a protocol for exposing tools/functions from remote servers.
"""

from __future__ import annotations

import uuid
from typing import Any, Optional

import httpx


# Module-level cache for MCP schemas
_mcp_schema_cache: dict[str, list[dict[str, Any]]] = {}


async def call_mcp_tool(
    call: dict[str, Any],
    *,
    url: str,
    token: Optional[str] = None,
) -> Any:
    """
    Call an MCP tool.

    Tool identity and tool arguments are carried as SEPARATE fields —
    never merged into one flat object. The previous flat-object
    contract ({**tool_args, "name": tool_name}) made a tool argument
    named ``name`` indistinguishable from the tool's own name, so
    tools with a top-level ``name`` parameter received None. Any
    unexpected top-level key raises so a legacy-shape caller fails
    loudly instead of silently dropping arguments.

    Args:
        call: {"name": str, "arguments": dict} — the tool's name and
            the model-provided arguments, forwarded verbatim.
        url: The MCP server URL.
        token: Optional bearer token for authentication.

    Returns:
        The result from the MCP tool call.

    Raises:
        RuntimeError: If the call fails.
    """
    if not url:
        raise RuntimeError("No MCP URL provided")

    tool_name = (call or {}).get("name")
    if not tool_name:
        raise RuntimeError("No tool name provided for MCP call")

    extra_keys = [k for k in call if k not in ("name", "arguments")]
    if extra_keys:
        raise RuntimeError(
            "call_mcp_tool takes {name, arguments} — unexpected top-level "
            f"keys: {', '.join(extra_keys)}. Tool arguments belong under "
            '"arguments".'
        )

    tool_args = call.get("arguments") or {}

    request_id = str(uuid.uuid4())

    payload = {
        "id": request_id,
        "method": "tools/call",
        "params": {
            "name": tool_name,
            "arguments": tool_args,
        },
    }

    headers: dict[str, str] = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"

    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(url, json=payload, headers=headers)
            response.raise_for_status()
            data = response.json()
            return data.get("result")

    except httpx.HTTPError as e:
        raise RuntimeError(f"Failed to call MCP tool \"{tool_name}\" at {url}: {e}") from e
    except Exception as e:
        raise RuntimeError(f"Failed to call MCP tool \"{tool_name}\" at {url}: {e}") from e


async def fetch_mcp_tools(
    url: str,
    token: Optional[str] = None,
    *,
    use_cache: bool = True,
) -> list[dict[str, Any]]:
    """
    Fetch all available tools from an MCP endpoint.

    Args:
        url: The MCP server URL.
        token: Optional bearer token for authentication.
        use_cache: Whether to use cached schemas (default True).

    Returns:
        List of tool schemas, each with name, description, and parameters.

    Raises:
        RuntimeError: If fetching fails.
    """
    if not url:
        raise RuntimeError("No MCP URL provided")

    # Check cache
    if use_cache and url in _mcp_schema_cache:
        return _mcp_schema_cache[url]

    request_id = str(uuid.uuid4())

    payload = {
        "id": request_id,
        "method": "tools/list",
        "params": {},
    }

    headers: dict[str, str] = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"

    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(url, json=payload, headers=headers)
            response.raise_for_status()
            data = response.json()

            tools_data = data.get("result", {}).get("tools", [])

            schemas: list[dict[str, Any]] = []
            for tool in tools_data:
                schema = {
                    "name": tool.get("name", ""),
                    "description": tool.get("description", ""),
                    "parameters": tool.get("inputSchema", {
                        "type": "object",
                        "properties": {},
                        "required": [],
                    }),
                }
                schemas.append(schema)

            # Cache the result
            _mcp_schema_cache[url] = schemas

            return schemas

    except httpx.HTTPError as e:
        raise RuntimeError(f"Failed to fetch MCP tools from {url}: {e}") from e
    except Exception as e:
        raise RuntimeError(f"Failed to fetch MCP tools from {url}: {e}") from e


def clear_mcp_cache() -> None:
    """Clear the MCP schema cache."""
    _mcp_schema_cache.clear()


def is_remote_mcp_tool(node: dict[str, Any]) -> bool:
    """
    Check if a node is a remote MCP tool.

    Args:
        node: The node to check.

    Returns:
        True if the node is a remote MCP tool.
    """
    return node.get("type") == "remote-mcp-tool"
